src/main/scala/analysis/SymbolicAccessAnalysis.scala

```diff
@@ -0,0 +1,152 @@
+package analysis
+
+import analysis.BitVectorEval.{bv2SignedInt, isNegative}
+import analysis.solvers.ForwardIDESolver
+import ir.IRWalk.procedure
+import ir.{Assign, BVADD, BinaryExpr, BitVecLiteral, BitVecType, CFGPosition, DirectCall, Expr, Extract, GoTo, IndirectCall, Literal, Memory, MemoryLoad, Procedure, Program, Register, Repeat, SignExtend, UnaryExpr, Variable, ZeroExtend}
+
+import java.math.BigInteger
+
+case class SymbolicAccess(accessor: Variable, symbolicBase: MemoryLocation, offset: BigInt) {
+  override def toString: String = s"SymbolicAccess($accessor, $symbolicBase, $offset)"
+}
+
+trait MemoryLocation {
+  val regionIdentifier: String
+
+  override def toString: String = s"MemoryRegion($regionIdentifier)"
+}
+
+case class StackLocation(override val regionIdentifier: String, proc: Procedure, size: BigInt) extends MemoryLocation {
+  override def toString: String = s"Stack($regionIdentifier,  $size)"
+}
+
+case class HeapLocation(override val regionIdentifier: String, proc: Procedure, size: BigInt) extends MemoryLocation {
+  override def toString: String = s"Heap($regionIdentifier, $size)"
+}
+
+case class DataLocation(override val regionIdentifier: String, start: BigInt, size: BigInt) extends MemoryLocation {
+  override def toString: String = s"Data($regionIdentifier, $start, $size)"
+}
+
+case class UnknownLocation(override val regionIdentifier: String, proc: Procedure) extends MemoryLocation {
+  override def toString: String = s"Unknown($regionIdentifier)"
+}
+
+/**
+ * environment transformers for SAA or symbolic access analysis
+ * Combination of reaching definitions and constant propagation
+ * elements in D are symbolic accesses of the form (variable, symbolic base, concrete offset)
+ * lattice L is a binary lattice with top and bottom
+ */
```

Comment:
What do the results of this analysis actually mean, exactly? It produces a Map[CFGPosition, Map[SymbolicAccess, TwoElement]] but I don't really follow what each part means here, especially the TwoElement.

Reply:
TwoElement is binary lattice element (top or bottom).
It's similar to live vars where there is a mapping from CFGPositions (flow-sensitive) to symbolic bases that are alive at that position.

Reply:
I know what a TwoElement is, what I'm wondering is what Top and Bottom represent in this context - what does it mean that a SymbolicAccess maps to Top or to Bottom?

Reply:
top is that definition of  Rx is valid (is alive)  and  bottom means that it no longer affects the environment (the definition is dead)

Reply:
Thanks, if you could add that as a comment that would be very helpful